### PR TITLE
CSD applied to Dansgaard-Oeschger events

### PR DESCRIPTION
**Pull request by JanJereczek**
This PR mainly provides an example (`docs/src/examples/do-events.jl`) of Critical Slowing Down applied on the $\delta^{18}O$ record from NGRIP. This time series displays DO-events and therefore many abrupt transitions in real-world, paleoclimatic data. Furthtermore, the same analysis is performed on simulation data from CLIMBER-X (less noisy and even time sampling), an Earth Model of Intermediate Complexity able to reproduce DO-like events.

Remark: The time steps  at which the DO-events are visible in the NGRIP timeseries come from Rasmussen 2014 and should be downloaded in future. For now, hand-coded. In the best case, this should be replaced by an automated procedure like those provided by `ChangePoints.jl`, which however did not give good results so far on the NGRIP data.

This PR additionally comes with minor suggestions:
- Use `DocumenterCitations.jl` as suggested and added `refs.bib`.
- The function `segmented_significance()` was introduced to perform the sliding window analysis such that a single significance value per segment and per indicator is output. This is subject to future changes since I want to introduce a new struct `GrowingWindowViewer`. We should discuss this live at some point.
- Replaced `docs/src/tutorial.md` by `docs/src/examples/tutorial.jl` such that the tutorial is built by Literate.jl.
- `precompute_ridgematrix` was renamed to `ridgematrix` since it simply computes the matrix.

### CODE REVIEW

**Comment by JanJereczek** [comment]
`estimate_indicator_changes` and `significant_transitions` now rely on multiple dispatch, by accepting `SlidingWindowConfig<:WindowedIndicatorResults` (renamed) or the newly introduced `SegmentWindowConfig<:WindowedIndicatorResults`. The latter requires the user to provide the start and end time of each segment into the config. To avoid any confusion about the functions performing the threaded loops, we now have two (internal) functions for this: `sliding_surrogates_loop!` and `segmented_surrogates_loop!`. For the sliding analysis, the `t_indicator, t_change` are simply vectors. For the segment analysis, they are each a vector of vectors: to me is the most intuitive way to handle the segmentation. Similar for `x_indicator`, which is a vector of matrices (dimensions of matrix as in in a sliding analysis). `x_change` is just a matrix with row index = segment index and column index = indicator index. Thus the user only computes one change metric value and therefore p-value per segment.

As suggested, we now use DocumenterCictations and the @cite macro.

Minor corrections were made to the DO-events example, especially the conclusions. It is still somewhat ugly that the transition times are hand-coded and we might want to make that more elegant in the future.

File: src/analysis/surrogates_significance.jl [hunk 1]
```diff
@@ -125,4 +124,52 @@ function indicator_metric_surrogates_loop!(
...
+Perform the surrogate significance analysis of the residual `r` over time `t` as defined by
+`indconfig` and `signif` for each segment `i` defined by `t_transitions[i]+margin[1] < t <
+t_transitions[i+1]+margin[2]`. For each segment, the window width to compute the change
+metric is chosen s.t. a single value of significance is returned (per indicator).
+"""
+function segmented_significance(t, r, t_transitions, margins, indconfig, change_metrics, signif)
+
+    if var(diff(t)) > 1e-8 * mean(diff(t))
+        error("Segment analysis only implemented for evenly-spaced time series.")
+    end
```
[the remark this answers is not in the document]
Comment (Datseris) [reply src/analysis/surrogates_significance.jl]: I mean define a function `is_evenly_sampled(x::AbstractVector) -> true/false` and use this function.

File: docs/src/examples/do-events.jl [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+## Preprocessing NGRIP
+
+Over this example, it will appear that the convenience of TransitionsInTimeseries leads the bulk of the code to be written for plotting and preprocessing. The latter consists in various steps $i = \lbrace 1,2,3 \rbrace$:
```
Comment (Datseris) [inline docs/src/examples/do-events.jl:8]: Here say something like "data pre-processing is not part of TransitionsInTimeseries.jl, but a step the user has to do before using the package. To present an example representing a complete scientific workflow, we will showcase typical data pre-processing here, that consist of the following steps:"

Then, after the numbered list, say "coincidentally, we will see how much convenience TransitionsInTimeseries.jl brings to the table. Almost all code in this example is dedicated to pre-processing, not to estimating transitions."

p.s.: Either use `TransitionsInTimeseries` (code font) or TransitionsInTimeseries.jl (with .jl).

File: docs/src/examples/do-events.jl [hunk 3]
```diff
@@ -0,0 +1,244 @@
+#=
+# Dansgaard-Oescher events
```
Comment (Datseris) [inline docs/src/examples/do-events.jl:2]:
```suggestion
# Dansgaard-Oescher events and Critical Slowing Down
```

This is a more representative title

File: docs/src/examples/do-events.jl [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+#=
+It here appears that not all transitions are preceeded by a significant increase of variance and AC1, even in the case of clean and evenly sampled time series. Let's check another case:
+=#
+
+figvec[2]
+
+#=
+Same here! Although CLIMBER-X does not represent real DO-events, the above-performed analysis might be hinting at the fact that not all DO transitions can be forecasted with CSD. Nonetheless, performing a CSD analysis can inform on the evolution of a system's resilience.
+
+# References
```
Comment (Datseris) [inline docs/src/examples/do-events.jl:240]: Move the references to a dedicated new top level page called "References" (like we have a tpp level page for Tutorial)

File: src/analysis/surrogates_significance-old.jl
Comment (Datseris) [inline src/analysis/surrogates_significance-old.jl:1]: Why does this file exist...?

File: docs/src/examples/tutorial.jl
Comment (Datseris) [inline docs/src/examples/tutorial.jl:1]: You moved the file, however the tutorial should NOT be under the examples. The tutorial is much more than an example and needs to be at the top level of the sidebar.  You need to modify how the tutorial is added to the `pages` in makedocs.jl to ensure this.

**Review by Datseris — CHANGES_REQUESTED** [review]
Hi Jan, thanks this is great! I am leaving one review, that is small and easy to tackle. I am short on time, so I'll do another review later, but you might as well address this one now if you can. 

I agree with all that you said. I also agree that Thus the user only computes one change metric value and therefore p-value per segment. But be sure to state this in the documentation string.

For the example that you have written: start segmenting the code blocks more. Don't have 100-200 l.o.c. in one code block, try to separate them more so that instead of 1 paragraph and then 100 lines of code, there is 1 sentence then 20 loc then another sentence then 20 loc etc.

That's not a fundamental rule or anything. Sometimes its fine to have 100 or 200 locs. But from time to time try to avoid it, especially at the early parts of the tutorial.

**Comment by Datseris** [comment]
> For the sliding analysis, the `t_indicator, t_change` are simply vectors. For the segment analysis, they are each a vector of vectors: to me is the most intuitive way to handle the segmentation. Similar for `x_indicator`, which is a vector of matrices (dimensions of matrix as in in a sliding analysis). `x_change` is just a matrix with row index = segment index and column index = indicator index. Thus the user only computes one change metric value and therefore p-value per segment

I did not understand that. Can you please try to explain again what you said above? Additionally, please point me to the specific lines of source code where you are talking about so that it is easier for me to follow.

File: docs/src/examples/do-events.jl (hunk 2 above)
Comment (JanJereczek) [reply docs/src/examples/do-events.jl:8]: Added the first suggestion, second one was already in there. Using TransitionsInTimeseries.jl.

File: src/analysis/surrogates_significance-old.jl
Comment (JanJereczek) [reply src/analysis/surrogates_significance-old.jl:1]: Why do we exist? No, just joking, file deleted! Thanks for spotting!

File: src/analysis/windowed_indicators.jl [hunk 5]
```diff
@@ -56,7 +68,92 @@ function WindowedIndicatorConfig(
...
+    change_metrics = map(f -> precompute(f, 1:T(width_cha)), change_metrics)
+
+    return SlidingWindowConfig(
+        indicators, change_metrics,
+        width_ind, stride_ind, width_cha, stride_cha, whichtime,
+    )
+end
+
+"""
+    SegmentWindowConfig(indicators, change_metrics; kwargs...) → config
```
[the remark this answers is not in the document]
Comment (JanJereczek) [reply src/analysis/windowed_indicators.jl:83]: Thanks for pointing out, very important in fact!!!

File: src/analysis/windowed_indicators.jl [hunk 6]
```diff
@@ -314,8 +325,31 @@ struct SlidingWindowResults{TT, T<:Real, X<:Real, XX<:AbstractVector{X},
...
+   with each column one indicator) of the `k`-th segment.
+- `t_indicator::Vector{Vector}`, with `t_indicator[k]` the time vector of the indicator
+  timeseries for the `k`-th segment.
+
+- `x_change::Matrix`, the change metric values with `x[k, i]` the change metric of the
+  `i`-th indicator for the `k`-th segment.
+- `t_change`, the time vector of the change metric.
+
+- [`config::SegmentWindowConfig`](@ref), used for the analysis.
+"""
```
Comment (JanJereczek) [inline src/analysis/windowed_indicators.jl:350]: Should clarify what I tried to convey when describing how the results of a segment analysis are structured!

File: src/analysis/surrogates_significance-old.jl [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            for i in 1:n_ind
+                indicator = config.indicators[i]
+                chametric = one2one ? config.change_metrics[i] : config.change_metrics[1]
+                z = view(x_indicator[k], :, i)
+                windowmap!(indicator, z, xseg;
+                    width = config.width_ind, stride = config.stride_ind
+                )
+                x_change[k, i] = chametric(z)
+            end
+        end
```
Comment (JanJereczek) [inline src/analysis/surrogates_significance-old.jl]: This is the main difference to a sliding window analysis

File: src/analysis/windowed_indicators.jl [hunk 8]
```diff
@@ -130,18 +215,90 @@ function estimate_indicator_changes(config::WindowedIndicatorConfig, x, t = each
...
 
     # put everything together in the output type
-    return WindowedIndicatorResults(
+    return SlidingWindowResults(
         t, x, t_indicator, x_indicator, t_change, x_change, config
     )
 end
 
+"""
+    estimate_indicator_changes(config::SegmentWindowConfig, x [,t]) → output
```
Comment (Datseris) [inline src/analysis/windowed_indicators.jl:224]: This function should NOT have two documentation strings. The function does _the same_ thing. It performs the estimation of indicators and change metrics as instructed by the configuration struct. Hence, we only need one simple docstring that shouldn't be more than 2 or 3 lines of text. The main explanation is happening in the configuration structs. The configuration strucs already say that they are passed into `estimate_indicator_cjhanges`.

**Review by Datseris — COMMENTED** [review]
Alright, seems good. There are some crucial documentation improvements we need to do first.

Besides those, I would argue that we should rename `SegmentWindowConfig` to `SegmentedWindowConfig`. The window segemenration has already been done by the user.

**Comment by Datseris** [comment]
@JanJereczek I merged this in. Notice that the PR violated the rule "do only 1 thing", as it actually did three things:

1. Expanded the API. With this PR we made it possible to add new different types for how to estimate change metrics and indicators.
2. Added a new way to estimate indicators and change metrics, the segmented window.
3. Added a new example.

However, that's okay for now, because all thse three things were tightly interlinked and we needed the example to see how the new interface could be used.

**Pull request merged**